AnimePahe isn't working

### Source information and language

AnimePahe 13.17

### Steps to reproduce

Streaming and downloading are not working

### Expected behavior

Should stream and download anime properly 

### Actual behavior

Showing - 
      Expected start of the object '{', but had 'EOF' instead JSON input: not enough arguments
     

### Aniyomi version

0.12.3.10

### Android version

Android 11

### Other details

_No response_

### Acknowledgements

- [X] I have searched the existing issues and this is a new ticket, **NOT** a duplicate or related to another open issue.
- [X] I have written a short but informative title.
- [X] I have updated the app to version **[0.13.5](https://github.com/jmir1/aniyomi/releases/latest)**.
- [X] I have updated all installed extensions.
- [X] I have tried the [troubleshooting guide](https://aniyomi.jmir.xyz/help/guides/troubleshooting/).
- [X] If this is an issue with the app itself, I should be opening an issue in the [app repository](https://github.com/jmir1/aniyomi/issues/new/choose).
- [X] I will fill out all of the requested information in this form.

Managed to grab a stack trace of the error:
```
02-14 18:52:57.222 28396 28540 E StandaloneCoroutine: Expected start of the object '{', but had 'EOF' instead
02-14 18:52:57.222 28396 28540 E StandaloneCoroutine: JSON input: not enough arguments
02-14 18:52:57.222 28396 28540 E StandaloneCoroutine: kotlinx.serialization.json.internal.JsonDecodingException: Expected start of the object '{', but had 'EOF' instead
02-14 18:52:57.222 28396 28540 E StandaloneCoroutine: JSON input: not enough arguments
02-14 18:52:57.222 28396 28540 E StandaloneCoroutine: 	at kotlinx.serialization.json.internal.JsonExceptionsKt.JsonDecodingException(JsonExceptions.kt:1)
02-14 18:52:57.222 28396 28540 E StandaloneCoroutine: 	at kotlinx.serialization.json.internal.JsonExceptionsKt.JsonDecodingException(JsonExceptions.kt:2)
02-14 18:52:57.222 28396 28540 E StandaloneCoroutine: 	at kotlinx.serialization.json.internal.AbstractJsonLexer.fail(AbstractJsonLexer.kt:1)
02-14 18:52:57.223 28396 28540 E StandaloneCoroutine: 	at kotlinx.serialization.json.internal.AbstractJsonLexer.fail$kotlinx_serialization_json(AbstractJsonLexer.kt:2)
02-14 18:52:57.223 28396 28540 E StandaloneCoroutine: 	at kotlinx.serialization.json.internal.AbstractJsonLexer.unexpectedToken(AbstractJsonLexer.kt:4)
02-14 18:52:57.223 28396 28540 E StandaloneCoroutine: 	at kotlinx.serialization.json.internal.StringJsonLexer.consumeNextToken(StringJsonLexer.kt:9)
02-14 18:52:57.223 28396 28540 E StandaloneCoroutine: 	at kotlinx.serialization.json.internal.StreamingJsonDecoder.beginStructure(StreamingJsonDecoder.kt:2)
02-14 18:52:57.223 28396 28540 E StandaloneCoroutine: 	at kotlinx.serialization.internal.AbstractCollectionSerializer.merge(CollectionSerializers.kt:3)
02-14 18:52:57.223 28396 28540 E StandaloneCoroutine: 	at kotlinx.serialization.internal.AbstractCollectionSerializer.deserialize(CollectionSerializers.kt:1)
02-14 18:52:57.223 28396 28540 E StandaloneCoroutine: 	at kotlinx.serialization.json.JsonObjectSerializer.deserialize(JsonElementSerializers.kt:3)
02-14 18:52:57.223 28396 28540 E StandaloneCoroutine: 	at kotlinx.serialization.json.JsonObjectSerializer.deserialize(JsonElementSerializers.kt:1)
02-14 18:52:57.223 28396 28540 E StandaloneCoroutine: 	at kotlinx.serialization.json.internal.PolymorphicKt.decodeSerializableValuePolymorphic(Polymorphic.kt:12)
02-14 18:52:57.223 28396 28540 E StandaloneCoroutine: 	at kotlinx.serialization.json.internal.StreamingJsonDecoder.decodeSerializableValue(StreamingJsonDecoder.kt:1)
02-14 18:52:57.223 28396 28540 E StandaloneCoroutine: 	at kotlinx.serialization.json.Json.decodeFromString(Json.kt:3)
02-14 18:52:57.223 28396 28540 E StandaloneCoroutine: 	at eu.kanade.tachiyomi.animeextension.en.animepahe.AnimePahe.videoListParse(AnimePahe.kt:337)
02-14 18:52:57.223 28396 28540 E StandaloneCoroutine: 	at eu.kanade.tachiyomi.ui.anime.AnimePresenter$$ExternalSyntheticLambda8.call(Unknown Source:46)
02-14 18:52:57.223 28396 28540 E StandaloneCoroutine: 	at rx.internal.operators.OnSubscribeMap$MapSubscriber.onNext(OnSubscribeMap.java:1)
02-14 18:52:57.223 28396 28540 E StandaloneCoroutine: 	at rx.internal.operators.OnSubscribeDoOnEach$DoOnEachSubscriber.onNext(OnSubscribeDoOnEach.java:3)
02-14 18:52:57.223 28396 28540 E StandaloneCoroutine: 	at eu.kanade.tachiyomi.network.OkHttpExtensionsKt$asObservable$1$requestArbiter$1.request(OkHttpExtensions.kt:4)
02-14 18:52:57.223 28396 28540 E StandaloneCoroutine: 	at rx.Subscriber.setProducer(Subscriber.java:8)
02-14 18:52:57.223 28396 28540 E StandaloneCoroutine: 	at rx.internal.operators.OnSubscribeMap$MapSubscriber.setProducer(OnSubscribeMap.java:1)
02-14 18:52:57.223 28396 28540 E StandaloneCoroutine: 	at rx.Subscriber.setProducer(Subscriber.java:6)
02-14 18:52:57.223 28396 28540 E StandaloneCoroutine: 	at eu.kanade.tachiyomi.network.OkHttpExtensionsKt$$ExternalSyntheticLambda0.call(Unknown Source:5)
02-14 18:52:57.223 28396 28540 E StandaloneCoroutine: 	at rx.Observable.unsafeSubscribe(Observable.java:2)
02-14 18:52:57.223 28396 28540 E StandaloneCoroutine: 	at rx.internal.operators.OnSubscribeDoOnEach.call(OnSubscribeDoOnEach.java:2)
02-14 18:52:57.223 28396 28540 E StandaloneCoroutine: 	at rx.internal.operators.OnSubscribeDoOnEach.call(OnSubscribeDoOnEach.java:1)
02-14 18:52:57.223 28396 28540 E StandaloneCoroutine: 	at rx.Observable.unsafeSubscribe(Observable.java:2)
02-14 18:52:57.223 28396 28540 E StandaloneCoroutine: 	at rx.internal.operators.OnSubscribeMap.call(OnSubscribeMap.java:4)
02-14 18:52:57.223 28396 28540 E StandaloneCoroutine: 	at rx.internal.operators.OnSubscribeMap.call(OnSubscribeMap.java:1)
02-14 18:52:57.223 28396 28540 E StandaloneCoroutine: 	at rx.Observable.unsafeSubscribe(Observable.java:2)
02-14 18:52:57.223 28396 28540 E StandaloneCoroutine: 	at rx.internal.operators.OnSubscribeFlattenIterable.call(OnSubscribeFlattenIterable.java:5)
02-14 18:52:57.223 28396 28540 E StandaloneCoroutine: 	at rx.internal.operators.OnSubscribeFlattenIterable.call(OnSubscribeFlattenIterable.java:1)
02-14 18:52:57.223 28396 28540 E StandaloneCoroutine: 	at rx.Observable.unsafeSubscribe(Observable.java:2)
02-14 18:52:57.223 28396 28540 E StandaloneCoroutine: 	at rx.internal.operators.OnSubscribeMap.call(OnSubscribeMap.java:4)
02-14 18:52:57.223 28396 28540 E StandaloneCoroutine: 	at rx.internal.operators.OnSubscribeMap.call(OnSubscribeMap.java:1)
02-14 18:52:57.223 28396 28540 E StandaloneCoroutine: 	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:4)
02-14 18:52:57.223 28396 28540 E StandaloneCoroutine: 	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:1)
02-14 18:52:57.223 28396 28540 E StandaloneCoroutine: 	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:4)
02-14 18:52:57.223 28396 28540 E StandaloneCoroutine: 	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:1)
02-14 18:52:57.223 28396 28540 E StandaloneCoroutine: 	at rx.Observable.unsafeSubscribe(Observable.java:2)
02-14 18:52:57.223 28396 28540 E StandaloneCoroutine: 	at rx.internal.operators.OnSubscribeDoOnEach.call(OnSubscribeDoOnEach.java:2)
02-14 18:52:57.223 28396 28540 E StandaloneCoroutine: 	at rx.internal.operators.OnSubscribeDoOnEach.call(OnSubscribeDoOnEach.java:1)
02-14 18:52:57.223 28396 28540 E StandaloneCoroutine: 	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:4)
02-14 18:52:57.223 28396 28540 E StandaloneCoroutine: 	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:1)
02-14 18:52:57.223 28396 28540 E StandaloneCoroutine: 	at rx.Observable.subscribe(Observable.java:25)
02-14 18:52:57.223 28396 28540 E StandaloneCoroutine: 	at eu.kanade.tachiyomi.util.lang.RxCoroutineBridgeKt.awaitOne(RxCoroutineBridge.kt:4)
02-14 18:52:57.223 28396 28540 E StandaloneCoroutine: 	at eu.kanade.tachiyomi.util.lang.RxCoroutineBridgeKt.awaitSingle(RxCoroutineBridge.kt:1)
02-14 18:52:57.223 28396 28540 E StandaloneCoroutine: 	at eu.kanade.tachiyomi.ui.player.PlayerActivity$awaitVideoList$1.invokeSuspend(PlayerActivity.kt:6)
02-14 18:52:57.223 28396 28540 E StandaloneCoroutine: 	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:4)
02-14 18:52:57.223 28396 28540 E StandaloneCoroutine: 	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:18)
02-14 18:52:57.223 28396 28540 E StandaloneCoroutine: 	at kotlinx.coroutines.internal.LimitedDispatcher.run(LimitedDispatcher.kt:2)
02-14 18:52:57.223 28396 28540 E StandaloneCoroutine: 	at kotlinx.coroutines.scheduling.TaskImpl.run(Tasks.kt:1)
02-14 18:52:57.223 28396 28540 E StandaloneCoroutine: 	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:1)
02-14 18:52:57.223 28396 28540 E StandaloneCoroutine: 	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:4)
02-14 18:52:57.223 28396 28540 E StandaloneCoroutine: 	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:4)
02-14 18:52:57.223 28396 28540 E StandaloneCoroutine: 	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:1)
02-14 18:52:57.223 28396 28540 E StandaloneCoroutine: Caused by: rx.exceptions.OnErrorThrowable$OnNextValue: OnError while emitting onNext value: okhttp3.Response.class
02-14 18:52:57.223 28396 28540 E StandaloneCoroutine: 	at rx.internal.operators.OnSubscribeMap$MapSubscriber.onNext(OnSubscribeMap.java:5)
02-14 18:52:57.223 28396 28540 E StandaloneCoroutine: 	... 39 more
02-14 18:52:57.328 28396 28540 E StandaloneCoroutine: java.lang.Exception
02-14 18:52:57.328 28396 28540 E StandaloneCoroutine: 	at eu.kanade.tachiyomi.extension.api.ExtensionGithubApi$findExtensions$2.invokeSuspend(ExtensionGithubApi.kt:19)
02-14 18:52:57.328 28396 28540 E StandaloneCoroutine: 	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:4)
02-14 18:52:57.328 28396 28540 E StandaloneCoroutine: 	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:18)
02-14 18:52:57.328 28396 28540 E StandaloneCoroutine: 	at kotlinx.coroutines.internal.LimitedDispatcher.run(LimitedDispatcher.kt:2)
02-14 18:52:57.328 28396 28540 E StandaloneCoroutine: 	at kotlinx.coroutines.scheduling.TaskImpl.run(Tasks.kt:1)
02-14 18:52:57.328 28396 28540 E StandaloneCoroutine: 	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:1)
02-14 18:52:57.328 28396 28540 E StandaloneCoroutine: 	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:4)
02-14 18:52:57.328 28396 28540 E StandaloneCoroutine: 	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:4)
02-14 18:52:57.328 28396 28540 E StandaloneCoroutine: 	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:1)
02-14 18:53:15.861 28396 28541 E StandaloneCoroutine: java.lang.Exception
02-14 18:53:15.861 28396 28541 E StandaloneCoroutine: 	at eu.kanade.tachiyomi.extension.api.ExtensionGithubApi$findExtensions$2.invokeSuspend(ExtensionGithubApi.kt:19)
02-14 18:53:15.861 28396 28541 E StandaloneCoroutine: 	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:4)
02-14 18:53:15.861 28396 28541 E StandaloneCoroutine: 	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:18)
02-14 18:53:15.861 28396 28541 E StandaloneCoroutine: 	at kotlinx.coroutines.internal.LimitedDispatcher.run(LimitedDispatcher.kt:2)
02-14 18:53:15.861 28396 28541 E StandaloneCoroutine: 	at kotlinx.coroutines.scheduling.TaskImpl.run(Tasks.kt:1)
02-14 18:53:15.861 28396 28541 E StandaloneCoroutine: 	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:1)
02-14 18:53:15.861 28396 28541 E StandaloneCoroutine: 	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:4)
02-14 18:53:15.861 28396 28541 E StandaloneCoroutine: 	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:4)
02-14 18:53:15.861 28396 28541 E StandaloneCoroutine: 	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:1)
```
Seems to suggest the issue is from this [parsing call](https://github.com/jmir1/aniyomi-extensions/blob/master/src/en/animepahe/src/eu/kanade/tachiyomi/animeextension/en/animepahe/AnimePahe.kt#L206).

--- a/src/en/animepahe/build.gradle
+++ b/src/en/animepahe/build.gradle
@@ -1,15 +1,17 @@
 apply plugin: 'com.android.application'
 apply plugin: 'kotlin-android'
+apply plugin: 'kotlinx-serialization'
 
 ext {
     extName = 'AnimePahe'
     pkgNameSuffix = 'en.animepahe'
     extClass = '.AnimePahe'
-    extVersionCode = 17
+    extVersionCode = 18
     libVersion = '13'
 }
 
 dependencies {
+    implementation "dev.datlag.jsunpacker:jsunpacker:1.0.1"
     ext.coroutinesVersion = "1.4.3"
     implementation("org.jetbrains.kotlinx:kotlinx-coroutines-core:$coroutinesVersion")
 }

--- a/src/en/animepahe/src/eu/kanade/tachiyomi/animeextension/en/animepahe/AnimePahe.kt
+++ b/src/en/animepahe/src/eu/kanade/tachiyomi/animeextension/en/animepahe/AnimePahe.kt
@@ -13,13 +13,15 @@ import eu.kanade.tachiyomi.animesource.model.SEpisode
 import eu.kanade.tachiyomi.animesource.model.Video
 import eu.kanade.tachiyomi.animesource.online.AnimeHttpSource
 import eu.kanade.tachiyomi.network.GET
+import eu.kanade.tachiyomi.network.asObservableSuccess
 import eu.kanade.tachiyomi.util.asJsoup
 import kotlinx.coroutines.Dispatchers
 import kotlinx.coroutines.runBlocking
 import kotlinx.coroutines.withContext
+import kotlinx.serialization.Serializable
 import kotlinx.serialization.decodeFromString
+import kotlinx.serialization.encodeToString
 import kotlinx.serialization.json.Json
-import kotlinx.serialization.json.JsonNull
 import kotlinx.serialization.json.JsonObject
 import kotlinx.serialization.json.float
 import kotlinx.serialization.json.int
@@ -30,6 +32,7 @@ import okhttp3.Headers
 import okhttp3.OkHttpClient
 import okhttp3.Request
 import okhttp3.Response
+import rx.Observable
 import uy.kohesive.injekt.Injekt
 import uy.kohesive.injekt.api.get
 import uy.kohesive.injekt.injectLazy
@@ -56,7 +59,8 @@ class AnimePahe : ConfigurableAnimeSource, AnimeHttpSource() {
     override val client: OkHttpClient = network.cloudflareClient
 
     override fun animeDetailsRequest(anime: SAnime): Request {
-        val animeId = anime.url.substringAfterLast("?anime_id=")
+        val parsed = json.decodeFromString<LinkData>(anime.url)
+        val animeId = parsed.url.substringAfterLast("?anime_id=")
         val session = getSession(anime.title, animeId)
         return GET("$baseUrl/anime/$session?anime_id=$animeId")
     }
@@ -111,7 +115,12 @@ class AnimePahe : ConfigurableAnimeSource, AnimeHttpSource() {
             val anime = SAnime.create()
             anime.title = item.jsonObject["title"]!!.jsonPrimitive.content
             val animeId = item.jsonObject["id"]!!.jsonPrimitive.int
-            anime.setUrlWithoutDomain("$baseUrl/anime/?anime_id=$animeId")
+            anime.setUrlWithoutDomain(
+                LinkData(
+                    "$baseUrl/anime/?anime_id=$animeId",
+                    item.jsonObject["session"]!!.jsonPrimitive.content
+                ).toJsonString()
+            )
             animeList.add(anime)
         }
         return AnimesPage(animeList, false)
@@ -136,7 +145,12 @@ class AnimePahe : ConfigurableAnimeSource, AnimeHttpSource() {
             val anime = SAnime.create()
             anime.title = item.jsonObject["anime_title"]!!.jsonPrimitive.content
             val animeId = item.jsonObject["anime_id"]!!.jsonPrimitive.int
-            anime.setUrlWithoutDomain("$baseUrl/anime/?anime_id=$animeId")
+            anime.setUrlWithoutDomain(
+                LinkData(
+                    "$baseUrl/anime/?anime_id=$animeId",
+                    item.jsonObject["anime_session"]!!.jsonPrimitive.content
+                ).toJsonString()
+            )
             anime.artist = item.jsonObject["fansub"]!!.jsonPrimitive.content
             animeList.add(anime)
         }
@@ -151,17 +165,33 @@ class AnimePahe : ConfigurableAnimeSource, AnimeHttpSource() {
         }
     }
 
+    override fun fetchEpisodeList(anime: SAnime): Observable<List<SEpisode>> {
+        val parsed = json.decodeFromString<LinkData>(anime.url)
+        return if (anime.status != SAnime.LICENSED) {
+            client.newCall(episodeListRequest(anime))
+                .asObservableSuccess()
+                .map { response ->
+                    episodeListParse(response, parsed.animeSession)
+                }
+        } else {
+            Observable.error(Exception("Licensed - No episodes to show"))
+        }
+    }
+
     override fun episodeListRequest(anime: SAnime): Request {
-        val animeId = anime.url.substringAfterLast("?anime_id=")
+        val parsed = json.decodeFromString<LinkData>(anime.url)
+        val animeId = parsed.url.substringAfterLast("?anime_id=")
         val session = getSession(anime.title, animeId)
         return GET("$baseUrl/api?m=release&id=$session&sort=episode_desc&page=1")
     }
 
-    override fun episodeListParse(response: Response): List<SEpisode> {
-        return recursivePages(response)
+    override fun episodeListParse(response: Response): List<SEpisode> = throw Exception("Not used")
+
+    private fun episodeListParse(response: Response, animeSession: String): List<SEpisode> {
+        return recursivePages(response, animeSession)
     }
 
-    private fun parseEpisodePage(jsonLine: String?): MutableList<SEpisode> {
+    private fun parseEpisodePage(jsonLine: String?, animeSession: String): MutableList<SEpisode> {
         val jsonData = jsonLine ?: return mutableListOf()
         val jObject = json.decodeFromString<JsonObject>(jsonData)
         val array = jObject["data"]!!.jsonArray
@@ -171,9 +201,8 @@ class AnimePahe : ConfigurableAnimeSource, AnimeHttpSource() {
             val episode = SEpisode.create()
             episode.date_upload = SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.US)
                 .parse(itemO["created_at"]!!.jsonPrimitive.content)!!.time
-            val animeId = itemO["anime_id"]!!.jsonPrimitive.int
             val session = itemO["session"]!!.jsonPrimitive.content
-            episode.setUrlWithoutDomain("$baseUrl/api?m=links&id=$animeId&session=$session&p=kwik")
+            episode.setUrlWithoutDomain("$baseUrl/play/$animeSession/$session")
             val epNum = itemO["episode"]!!.jsonPrimitive.float
             episode.episode_number = epNum
             val epNumString = if (epNum % 1F == 0F) epNum.toInt().toString() else epNum.toString()
@@ -183,16 +212,16 @@ class AnimePahe : ConfigurableAnimeSource, AnimeHttpSource() {
         return episodeList
     }
 
-    private fun recursivePages(response: Response): List<SEpisode> {
+    private fun recursivePages(response: Response, animeSession: String): List<SEpisode> {
         val responseString = response.body!!.string()
         val jObject = json.decodeFromString<JsonObject>(responseString)
         val lastPage = jObject["last_page"]!!.jsonPrimitive.int
         val page = jObject["current_page"]!!.jsonPrimitive.int
         val hasNextPage = page < lastPage
-        val returnList = parseEpisodePage(responseString)
+        val returnList = parseEpisodePage(responseString, animeSession)
         if (hasNextPage) {
             val nextPage = nextPageRequest(response.request.url.toString(), page + 1)
-            returnList += recursivePages(nextPage)
+            returnList += recursivePages(nextPage, animeSession)
         }
         return returnList
     }
@@ -203,18 +232,17 @@ class AnimePahe : ConfigurableAnimeSource, AnimeHttpSource() {
     }
 
     override fun videoListParse(response: Response): List<Video> {
-        val array = json.decodeFromString<JsonObject>(response.body!!.string())
-            .jsonObject["data"]!!.jsonArray
-        val videos = mutableListOf<Video>()
-        for (item in array) {
-            val quality = item.jsonObject.keys.first()
-            val paheWinLink = item.jsonObject[quality]!!.jsonObject["kwik_pahewin"]!!.jsonPrimitive.content
-            val kwikLink = item.jsonObject[quality]!!.jsonObject["kwik"]!!.jsonPrimitive.content
-            val audio = item.jsonObject[quality]!!.jsonObject["audio"]!!
-            val qualityString = if (audio is JsonNull) "${quality}p" else "${quality}p (" + audio.jsonPrimitive.content + " audio)"
-            videos.add(getVideo(paheWinLink, kwikLink, qualityString))
+        val document = response.asJsoup()
+        val videoList = mutableListOf<Video>()
+
+        document.select("div#resolutionMenu > button").forEachIndexed { index, btn ->
+            val kwikLink = btn.attr("data-src")
+            val quality = btn.text()
+            val paheWinLink = document.select("div#pickDownload > a")[index].attr("href")
+            videoList.add(getVideo(paheWinLink, kwikLink, quality))
         }
-        return videos
+
+        return videoList
     }
 
     private fun getVideo(paheUrl: String, kwikUrl: String, quality: String): Video {
@@ -233,27 +261,14 @@ class AnimePahe : ConfigurableAnimeSource, AnimeHttpSource() {
     override fun List<Video>.sort(): List<Video> {
         val subPreference = preferences.getString("preferred_sub", "jpn")!!
         val quality = preferences.getString("preferred_quality", "1080")!!
-        val qualityList = mutableListOf<Video>()
-        val newList = mutableListOf<Video>()
-        var preferred = 0
-        for (video in this.reversed()) {
-            if (video.quality.contains(quality)) {
-                qualityList.add(preferred, video)
-                preferred++
-            } else {
-                qualityList.add(video)
-            }
-        }
-        preferred = 0
-        for (video in qualityList) {
-            if (video.quality.contains(subPreference)) {
-                newList.add(preferred, video)
-                preferred++
-            } else {
-                newList.add(video)
-            }
-        }
-        return newList
+        val shouldEndWithEng = (subPreference == "eng")
+
+        return this.sortedWith(
+            compareBy(
+                { it.quality.contains(quality) },
+                { it.quality.endsWith("eng", true) == shouldEndWithEng },
+            )
+        ).reversed()
     }
 
     override fun setupPreferenceScreen(screen: PreferenceScreen) {
@@ -318,5 +333,15 @@ class AnimePahe : ConfigurableAnimeSource, AnimeHttpSource() {
         screen.addPreference(domainPref)
         screen.addPreference(subPref)
         screen.addPreference(linkPref)
+    }
+
+    @Serializable
+    data class LinkData(
+        val url: String,
+        val animeSession: String
+    )
+
+    private fun LinkData.toJsonString(): String {
+        return json.encodeToString(this)
     }
 }

--- a/src/en/animepahe/src/eu/kanade/tachiyomi/animeextension/en/animepahe/KwikExtractor.kt
+++ b/src/en/animepahe/src/eu/kanade/tachiyomi/animeextension/en/animepahe/KwikExtractor.kt
@@ -27,8 +27,10 @@ SOFTWARE.
 
 package eu.kanade.tachiyomi.animeextension.en.animepahe
 
+import dev.datlag.jsunpacker.JsUnpacker
 import eu.kanade.tachiyomi.network.GET
 import eu.kanade.tachiyomi.network.POST
+import eu.kanade.tachiyomi.util.asJsoup
 import okhttp3.FormBody
 import okhttp3.Headers
 import okhttp3.OkHttpClient
@@ -48,13 +50,10 @@ class KwikExtractor(private val client: OkHttpClient) {
 
     fun getHlsStreamUrl(kwikUrl: String, referer: String): String {
         val eContent = client.newCall(GET(kwikUrl, Headers.headersOf("referer", referer)))
-            .execute().body!!.string()
-        val substring = eContent.substringAfterLast("m3u8|uwu|").substringBefore("'")
-        val urlParts = substring.split("|").reversed()
-        assert(urlParts.lastIndex == 8)
-        return urlParts[0] + "://" + urlParts[1] + "-" + urlParts[2] + "." + urlParts[3] + "." +
-            urlParts[4] + "." + urlParts[5] + "/" + urlParts[6] + "/" + urlParts[7] + "/" +
-            urlParts[8] + "/uwu.m3u8"
+            .execute().asJsoup()
+        val script = eContent.selectFirst("script:containsData(eval\\(function)").data().substringAfterLast("eval(function(")
+        val unpacked = JsUnpacker.unpackAndCombine("eval(function($script")!!
+        return unpacked.substringAfter("const source=\\'").substringBefore("\\';")
     }
 
     fun getStreamUrlFromKwik(paheUrl: String): String {

